### PR TITLE
feat: Add state and the ability to set the nick and status message

### PR DESCRIPTION
**Pull request by JFreegman**
This change is [Reviewable](https://reviewable.io/reviews/toktok/btox/19)

### CODE REVIEW

File: lib/appstate.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import 'package:flutter/material.dart';
+
+class AppState {
+  var nickName = ValueNotifier<String>('Yanciman');
+  var userStatus = ValueNotifier<String>('Producing works of art in Kannywood');
+
+  void setNick(String nickName) {
+    this.nickName = ValueNotifier<String>(nickName);
```
Comment (robinlinden) [inline lib/appstate.dart:8]: This should probably be setting `this.nickName.value` and not creating a new `ValueNotifier<String>`? Assigning to value uses the [setter](https://dart.dev/guides/language/language-tour#getters-and-setters) magic to run some code during the assignment, in this case notifying all listeners.

This also allows both members to be final.

File: lib/appstate.dart [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import 'package:flutter/material.dart';
+
+class AppState {
+  var nickName = ValueNotifier<String>('Yanciman');
```
Comment (robinlinden) [inline lib/appstate.dart:4]: You could implement setters for these, so `state.nickName = 'robinlinguini';` will set the value of the ValueNotifier?

Not sure if it's too magic, but it seems to be how things are done in Flutter.

See: https://dart.dev/guides/language/language-tour#getters-and-setters

File: lib/contact_list_page.dart [hunk 3]
```diff
@@ -52,27 +55,35 @@ class _ContactListPageState extends State<ContactListPage> {
...
-                    fontSize: 12.0,
-                    color: Colors.white,
-                    fontWeight: FontWeight.w100,
-                  ),
-                ),
+                title: ValueListenableBuilder(
+                    valueListenable: widget.appState.nickName,
+                    builder: (context, nickName, _) {
+                      return Text(
+                        widget.appState.getNick(),
```
Comment (robinlinden) [inline lib/contact_list_page.dart:67]: This should probably be using the `nickName` the builder notifies you of? No need to use (or even have a getter) if we do that. If we drop the `String getX()` methods from the AppState class, we also make sure that no code in the app can accidentally miss setting up a subscription and end up with stale data.

File: lib/profile.dart [hunk 4]
```diff
@@ -1,23 +1,145 @@
 import 'package:flutter/material.dart';
 
+import 'appstate.dart';
+import 'strings.dart';
+
+// ignore: must_be_immutable
```
Comment (robinlinden) [inline lib/profile.dart:6]: Why not mark the AppState final and const the constructor?

File: lib/profile.dart [hunk 5]
```diff
@@ -1,23 +1,145 @@
...
+    return _nickInputController.text.isNotEmpty &&
+        _nickInputController.text.length <= 32;
+  }
+
+  bool _statusMessageIsValid(String nick) {
+    return _statusMessageInputController.text.length <= 256;
+  }
+
+  _onValidate() {
+    if (!_nickIsValid(_nickInputController.text) ||
```
Comment (robinlinden) [inline lib/profile.dart:44]: You can use `_formKey.currentState.validate()` to run every validation function registered in the form. That way you don't need any helpers for validation anymore.

File: lib/profile.dart [hunk 6]
```diff
@@ -1,23 +1,145 @@
...
+        title: const Text(Strings.menuProfile),
       ),
-      body: const Center(
-        child: Text('Username'),
+      body: Form(
+        key: _formKey,
+        child: Column(
+          children: [
+            const Padding(
+              padding: EdgeInsets.all(10),
```
Comment (robinlinden) [inline lib/profile.dart:78]: I think in general, all UI distances in Flutter are powers of two. I know e.g. the ListTile uses 16 as its default padding.

File: lib/profile.dart [hunk 7]
```diff
@@ -1,23 +1,145 @@
...
+            ),
+            const Text(
+              'Nickname',
```
Comment (robinlinden) [inline lib/profile.dart:81]: Pull these strings out into strings.dart as well?

File: lib/profile.dart [hunk 8]
```diff
@@ -1,23 +1,145 @@
...
+                validator: (value) {
+                  value ??= '';
+                  if (!_nickIsValid(value)) {
+                    return Strings.nickLengthError;
+                  }
+
+                  return null;
+                },
+                controller: _nickInputController,
+                textInputAction: TextInputAction.send,
```
Comment (robinlinden) [inline lib/profile.dart:101]: If you change these to `TextInputAction.next`, it'll move to the next field on enter, and submit the form once you hit enter on the last form field before the button.

File: lib/profile.dart [hunk 9]
```diff
@@ -1,23 +1,145 @@
...
+              'Nickname',
+              style: TextStyle(
+                fontSize: 16.0,
+                fontWeight: FontWeight.bold,
+              ),
+            ),
+            Padding(
+              padding: const EdgeInsets.all(10),
+              child: TextFormField(
+                key: const Key('nickName'),
```
Comment (robinlinden) [inline lib/profile.dart:90]: I had a key for my text field so I could find it in the unit tests, so these aren't needed right now.

File: lib/profile.dart [hunk 10]
```diff
@@ -1,23 +1,145 @@
...
+                    return Strings.statusMessageLengthError;
+                  }
+
+                  return null;
+                },
+                controller: _statusMessageInputController,
+                textInputAction: TextInputAction.send,
+              ),
+            ),
+            ElevatedButton(
```
Comment (robinlinden) [inline lib/profile.dart:131]: The dynamic bits in this are really neat!

**Review by robinlinden — CHANGES_REQUESTED** [review]

File: lib/appstate.dart (hunk 2 above)
Comment (JFreegman) [reply lib/appstate.dart]: Magic indeed

File: lib/profile.dart (hunk 5 above)
Comment (JFreegman) [reply lib/profile.dart:50]: The onValidate method is what pushes the button and sets the state. We don't want to call that every keystroke. Unless I'm missing something...

**Review by robinlinden — APPROVED** [review]

**Pull request merged**